app/models/concerns/operations/stock_operations.rb

```diff
@@ -68,5 +70,98 @@ def register_loss(package, quantity:, from_location:)
         location: from_location
       )
     end
+
+    def pack_or_unpack(params, user_id)
+      raise ActionNotAllowedError.new unless PACK_UNPACK_ALLOWED_ACTIONS.include?(params[:task])
+      PackUnpack.new(params, user_id).public_send(params[:task])
+    end
+
+    class PackUnpack
+      def initialize(params, user_id)
+        @cause = Package.find(params[:id]) # box or pallet
+        @item = Package.find(params[:item_id]) # item to add or remove
+        @user_id = user_id
+        @location_id = params[:location_id]
+        @quantity = params[:quantity].to_i # quantity to pack or unpack
+      end
+
+      def pack
+        return error(I18n.t("box_pallet.errors.adding_box_to_box")) if adding_box_to_a_box?
+        return error(I18n.t("box_pallet.errors.item_designated")) if item_designated?
+        return error(I18n.t("box_pallet.errors.invalid_quantity")) if invalid_quantity?
+        pkg_inventory = pack_or_unpack(PackagesInventory::Actions::PACK)
+        response(pkg_inventory)
+      end
+
+      def unpack
+        pkg_inventory = pack_or_unpack(PackagesInventory::Actions::UNPACK)
+        response(pkg_inventory)
+      end
+
+      private
+
+      def pack_or_unpack(task)
+        return unless @quantity.positive?
+        PackagesInventory.new(
+          package: @item,
+          source: @cause,
+          action: task,
+          location_id: @location_id,
+          user_id: @user_id,
+          quantity: quantity(task)
+        )
+      end
+
+      def quantity(task)
+        task.eql?("pack") ? @quantity * -1 : @quantity
+      end
+
+      def error(error)
+        { errors: [error], success: false }
+      end
+
+      def response(pkg_inventory)
+        return unless pkg_inventory
+        if pkg_inventory.save
+          { packages_inventory: pkg_inventory, success: true }
+        elsif pkg_inventory.errors
+          error(pkg_inventory.errors.full_messages)
+        end
+      end
+
+      def invalid_quantity?
+        available_quantity = available_quantity_on_location(@location_id)
+        @quantity > available_quantity
+      end
+
+      def available_quantity_on_location(location_id)
+        PackagesLocation.where(location_id: location_id, package_id: @item.id).first.quantity
+      end
+
+      def adding_box_to_a_box?
+        @item.box? && @cause.box?
+      end
+
+      def item_designated?
+        @item.order.presence
+      end
+    end
+
+    # --- Exceptions
+
+    class OperationsError < StandardError; end
```

Comment:
All the errors have been moved to `lib/goodcity/errors.rb` , you might want to move those there : )

Reply:
sure @patrixr , I'll move those errors to that class.